### PR TITLE
Avoid unused greedy normalization and LFM vocabulary projection

### PR DESCRIPTION
**Pull request by Blaizzy**
Greedy generation without requested probabilities still normalizes the full vocabulary, and LFM prefill projects unused hidden positions into vocabulary logits. Skip those operations while preserving sampling, requested logprobs, and full cache/hidden-state processing.

**Depends on #1630.** Base branch: `pc/scheduler-decode-emit-prefill`. Merge #1630 first. This diff contains only four files of model/sampling changes and their tests.

- Share a greedy fast path between prefill and decode. Normalize when sampling or returning token/top probabilities, including when logits processors are present.
- Request the final intermediate prefill logit and honor `logits_to_keep` in LFM before vocabulary projection.
- Verify both tied and untied output heads retain full hidden states and caches.

### Validation

- The updated #1630 base passed its full Python 3.10 [GitHub CI run](https://github.com/Blaizzy/mlx-vlm/actions/runs/35628407929). The complete stack is covered by the full local suite below.
- **Full local suite: 1,915 passed, 4 skipped** on the complete stack (Python 3.12), including multimodal, audio, video, image generation, extraction, training, and processor tests.

- Coverage includes sampling normalization, logits processors, requested probabilities, intermediate logits hints, and LFM cache/hidden-state preservation.
- Black, isort, autoflake, compileall, and `git diff --check` passed.
- Compare `#1630` with `#1630 + #1795` to isolate this PR's contribution. No general throughput speedup is claimed.

### Tradeoffs

- The normalization shortcut applies only to greedy decoding without requested probabilities. Sampling and probability reporting retain normalization; the vocabulary-projection change applies to LFM.
- End-to-end throughput changes are small and workload-dependent in the measurements below. Scheduler latency/memory gains come from #1630; they should not be attributed to this optimization alone.
- Three repeats on two text workloads do not establish a general speedup or output-quality equivalence across batch schedules and model families.

### Measurements

Apple M5 Max / 128 GB, MLX 0.32.2, LiquidAI/LFM2.5-2.6B BF16; a 401-token lead request starts first and peers arrive after its first token. Each request emits 64 greedy tokens. Prefill setting: 2048; APC and speculative decoding off. Each case/version has one warmup and three measured trials with balanced rotating order; only one variant generates at a time. Server tokenizer, queue, GPU worker, and output delivery are included; HTTP overhead is excluded. Desktop applications remained active; no outliers were excluded. All 24 trials completed (18 measured and 6 warmups), with matching prompt hashes and dependencies and verified output lengths.

Values are medians; maximum/P95 token gaps are computed per trial before summarizing. Three repeats on two synthetic text workloads do not establish a general throughput speedup or output-quality equivalence across models and batch shapes.

| Arriving peers | Version | Output tok/s | Max token gap (ms) | P95 gap (ms) | Peak GB |
|---|---|---:|---:|---:|---:|
| 7 × 2K | Main | 188.00 | 1594.4 | 17.1 | 8.31 |
| 7 × 2K | #1630 | 198.70 | 237.5 | 16.5 | 6.68 |
| 7 × 2K | #1630 + #1795 | 191.44 | 243.5 | 17.5 | 6.68 |
| 1 × 16K | Main | 40.75 | 284.7 | 236.4 | 6.62 |
| 1 × 16K | #1630 | 40.30 | 286.7 | 238.5 | 6.62 |
| 1 × 16K | #1630 + #1795 | 40.94 | 295.3 | 235.3 | 6.62 |

Matched comparisons (distinct from ratios of table medians):

- 7 × 2K, Main → #1630: matched throughput median -0.3% (range -3.7% to +5.7%); identical output token multisets in 0/3 pairs.
- 7 × 2K, #1630 → #1630 + #1795: matched throughput median -0.4% (range -3.7% to -0.1%); identical output token multisets in 2/3 pairs.
- 1 × 16K, Main → #1630: matched throughput median -1.1% (range -2.3% to +1.5%); identical output token multisets in 3/3 pairs.
- 1 × 16K, #1630 → #1630 + #1795: matched throughput median +2.7% (range -3.1% to +3.4%); identical output token multisets in 3/3 pairs.

One 7 × 2K full-stack trial delivered peer first tokens in two timing groups (about 1.565s and 1.824s) and produced a third distinct output sequence; it is retained in all summaries. These results do not establish bitwise output equivalence across changing schedules.

Measured revisions: Main `3f09d925`, #1630 `1faf3db6`, #1630 + #1795 `127b65ee`.

### CODE REVIEW

**Comment by Blaizzy** [comment]
## 16K batch scaling follow-up (B1–B16)

Same M3 Ultra / 512 GB host and BF16 LFM2.5-2.6B model as the PR benchmark: exact 16,384-token prompts, 64 greedy output tokens/request, APC off, 2,048-token prefill chunks, fresh server per batch size, one warmup, median of three measured trials.

### Fully admitted decode batches

This uses the existing `MLX_VLM_BATCH_COALESCE_MS=350` knob so all serialized 16K requests reach the idle queue before prefill starts.

| Batch | Aggregate decode | vs B1 | Median/request | Wall | Peak memory | Rows exact |
| ---: | ---: | ---: | ---: | ---: | ---: | :---: |
| 1 | 96.18 tok/s | 1.000x | 96.18 tok/s | 4.843 s | 6.98 GB | yes |
| 2 | 101.78 tok/s | 1.058x | 50.89 tok/s | 9.235 s | 7.50 GB | yes |
| 4 | 195.90 tok/s | 2.037x | 48.98 tok/s | 16.894 s | 10.12 GB | yes |
| 8 | 356.81 tok/s | 3.710x | 44.60 tok/s | 32.341 s | 19.28 GB | yes |
| 16 | **711.34 tok/s** | **7.396x** | 44.48 tok/s | 64.663 s | 19.28 GB | yes |

Decode scales to B16 without OOM. End-to-end generated-token throughput rises from 13.21 tok/s at B1 to 15.84 tok/s at B16 because wall time is dominated by the intentionally singleton LFM prompt prefills.

Repro using the exact benchmark script already embedded in the PR description:

```bash
uv run python /tmp/lfm16k_batch2_focus.py \
  --model /path/to/LFM2.5-2.6B-bf16 \
  --batches 1 2 4 8 16 \
  --warmups 1 --trials 3 --allow-mismatch \
  --coalesce-ms 350
```

### Default 20 ms admission window

| Batch | Aggregate decode | Median/request | Wall | Peak memory | Rows exact |
| ---: | ---: | ---: | ---: | ---: | :---: |
| 1 | 96.62 tok/s | 96.62 tok/s | 4.499 s | 6.98 GB | yes |
| 2 | 101.87 tok/s | 50.93 tok/s | 8.880 s | 7.50 GB | yes |
| 4 | 28.43 tok/s | 10.22 tok/s | 16.801 s | 10.49 GB | yes |
| 8 | 20.87 tok/s | 4.45 tok/s | 32.899 s | 20.62 GB | yes |
| 16 | 36.29 tok/s | 2.82 tok/s | 65.411 s | 20.62 GB | **no** |

The default remains tuned for B2. At B4+, serialized 16K tokenization outlasts the 20 ms window; late requests join after decode has begun, and their singleton prefills interrupt earlier rows. At B16, measured runs produced 2–3 text variants, all first diverging at the known token-18 BF16 near-tie. Increasing the existing coalescing knob admits the entire group up front, removes the prefill/decode interference, and restores exact agreement within every batch.

### B16 full-logit correctness

Using singleton-prefilled caches and an identical canonical history for 64 decode positions:

- first token: B1 `[597]`; B16 `[597] * 16`
- B16 row-to-row maximum absolute difference: **0.0**
- worst B1-versus-B16 full-vocabulary absolute difference: **0.320312** (asserted BF16 tolerance: 0.5)
- argmax differences only at positions 18 and 44, where candidates are tied or one BF16 step apart—the same positions observed at B2

<details>
<summary>Exact B16 correctness script</summary>

```python
import argparse

import mlx.core as mx
import numpy as np
from transformers import AutoTokenizer

from mlx_vlm.generate import PromptProcessingBatch
from mlx_vlm.utils import load


CONTEXT = 16_384
STEPS = 64
SUFFIX = "\nGenerate an endless sequence of numbers separated by spaces. Begin now:"


def prompt_ids(tokenizer, content):
    encoded = tokenizer.apply_chat_template(
        [{"role": "user", "content": content}],
        tokenize=True,
        add_generation_prompt=True,
    )
    return encoded["input_ids"] if hasattr(encoded, "keys") else encoded


def exact_ids(tokenizer):
    repeats = CONTEXT - 24
    for _ in range(8):
        ids = prompt_ids(tokenizer, "x " * repeats + SUFFIX)
        if len(ids) == CONTEXT:
            return ids
        repeats += CONTEXT - len(ids)
    raise RuntimeError(f"exact prompt failed: {len(ids)}")


def singleton_prefill(model, processor, ids, uid):
    input_ids = mx.array([ids], dtype=mx.int32)
    inputs_embeds = model.get_input_embeddings(input_ids).inputs_embeds
    batch = PromptProcessingBatch(
        model=model.language_model,
        uids=[uid],
        input_ids=[ids],
        max_tokens=[STEPS],
        inputs_embeds=inputs_embeds,
        prompt_kwargs={},
        prefill_step_size=2048,
        greedy_sampling=True,
    )
    while batch.needs_processing():
        batch.prompt_step()
    generated = batch.generate(
        lambda logprobs: mx.argmax(logprobs, axis=-1),
        processor.stopping_criteria,
        compute_logprobs=False,
    )
    return generated


def top_two(values):
    indices = np.argpartition(values, -2)[-2:]
    indices = indices[np.argsort(values[indices])[::-1]]
    return [(int(index), float(values[index])) for index in indices]


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--model", required=True)
    parser.add_argument("--batch-size", type=int, default=2)
    args = parser.parse_args()

    tokenizer = AutoTokenizer.from_pretrained(args.model)
    ids = exact_ids(tokenizer)
    model, processor = load(args.model)
    single = singleton_prefill(model, processor, ids, 0)
    rows = [
        singleton_prefill(model, processor, ids, uid)
        for uid in range(1, args.batch_size + 1)
    ]
    batched = rows[0]
    for row in rows[1:]:
        batched.extend(row)

    current_single = single._next_tokens
    current_batch = batched._next_tokens
    mx.eval(current_single, current_batch)
    print(
        f"first_tokens single={current_single.tolist()} "
        f"batch={current_batch.tolist()}",
        flush=True,
    )
    assert current_batch.tolist() == current_single.tolist() * args.batch_size

    worst = 0.0
    worst_row_delta = 0.0
    mismatch_positions = []
    for position in range(1, STEPS):
        out_single = model.language_model(
            current_single[:, None], cache=single.prompt_cache
        )
        out_batch = model.language_model(
            current_batch[:, None], cache=batched.prompt_cache
        )
        raw_logits_single = out_single.logits[:, -1, :]
        raw_logits_batch = out_batch.logits[:, -1, :]
        logits_single = raw_logits_single.astype(mx.float32)
        logits_batch = raw_logits_batch.astype(mx.float32)
        mlx_single_argmax = mx.argmax(logits_single, axis=-1)
        mlx_batch_argmax = mx.argmax(logits_batch, axis=-1)
        mx.eval(
            logits_single,
            logits_batch,
            mlx_single_argmax,
            mlx_batch_argmax,
        )
        one = np.asarray(logits_single[0])
        two = np.asarray(logits_batch[0])
        row_delta = float(
            np.max(
                np.abs(
                    np.asarray(logits_batch)
                    - np.asarray(logits_batch[0])[None, :]
                )
            )
        )
        max_abs = float(np.max(np.abs(one - two)))
        worst = max(worst, max_abs)
        worst_row_delta = max(worst_row_delta, row_delta)
        top_single = top_two(one)
        top_batch = top_two(two)
        margin = top_single[0][1] - top_single[1][1]
        single_argmax = mlx_single_argmax.tolist()[0]
        batch_argmax = mlx_batch_argmax.tolist()[0]
        mismatch = single_argmax != batch_argmax
        if mismatch:
            mismatch_positions.append(position)
        reduction_mismatch = (
            mlx_single_argmax.tolist()[0] != top_single[0][0]
            or mlx_batch_argmax.tolist()[0] != top_batch[0][0]
        )
        if mismatch or reduction_mismatch or position <= 3 or position % 8 == 0 or position == 18:
            print(
                f"position={position} mismatch={mismatch} "
                f"reduction_mismatch={reduction_mismatch} "
                f"mlx_single={mlx_single_argmax.tolist()} "
                f"mlx_batch={mlx_batch_argmax.tolist()} "
                f"max_abs={max_abs:.6f} "
                f"row_delta={row_delta:.6f} margin={margin:.6f} "
                f"single={top_single} batch={top_batch}",
                flush=True,
            )
        canonical = single_argmax
        current_single = mx.array([canonical], dtype=mx.int32)
        current_batch = mx.array(
            [canonical] * args.batch_size, dtype=mx.int32
        )

    assert worst <= 0.5, worst
    assert worst_row_delta == 0.0, worst_row_delta
    print(
        f"CORRECTNESS worst_max_abs={worst:.6f} "
        f"batch_row_max_abs={worst_row_delta:.6f} "
        f"argmax_tie_positions={mismatch_positions}",
        flush=True,
    )


if __name__ == "__main__":
    main()

```

</details>
